[DOCS-221] Best Pratices: Input Validation (#154)
cobalthq/cobalt-product-public-docs: content/en/BestPractices/akhan_input_validation.md

website that allows me to buy a negative quantity of products."

Review this code sample from the [OWASP Juice Shop](https://github.com/juice-shop/juice-shop/).
It's a perfect example of how an innocent mistake can cause huge business losses:	

```ts
 async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
  const product = await QuantityModel.findOne({ where: { ProductId: id } })
  if (!product) {
    throw new Error('No such product found!')
  }

  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
    if (product.quantity >= quantity) { // enough in stock?
      next()
    } else {
      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
    }
  } else {
    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
  }
}
```

### Analysis

The OWASP Juice Shop uses an [API
Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)

> to call a

Reviewer: The OWASP Juice Shop frontend initiates an API POST request to add items to the user's basket. The middleware function that is mounted on that API route handler (or API path) calls the `quantityCheck()` function.

Author: @sourcecodereviewer I hope my suggested change[ in this comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945674503) matches what you're asking for

> The OWASP Juice Shop uses an [API
> Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)
> to call a
> [`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
> function.

Author: ```suggestion
The OWASP Juice Shop frontend initiates an [API
Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370) request to add items
to the user's basket. The middleware function that is mounted on that API route handler (or API path) calls the
[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
function.
```

It checks the quantity in the request payload with validation logic, using

> `product.limitPerUser`. However, it does not check if the quantity exceeds the:

Reviewer: The `quantityCheck()` function checks if the quantity is valid and doesn't exceed the limit allowed per user or, the limit allowed per order. It bypasses the check if the user is a `deluxe` user.

Author: See [next comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945688933) for response

> It checks the quantity in the request payload with validation logic, using
> `product.limitPerUser`. However, it does not check if the quantity exceeds the:
> 
> - Limit per user
> - Limit per order
> 
> It even skips the check for deluxe users (`security.isDeluxe(req)`). However, it
> does *not* check if the quantity is positive.

Author: @sourcecodereviewer How does this look? 

```suggestion
The `quantityCheck()` function validates the quantity in the request payload, as it checks:

- If the product is available (`const product = await QuantityModel.findOne({ where: { ProductId: id } })`)
- If the quantity exceeds the quantity available (`product.quantity >= quantity`)
- If the quantity exceeds the limit per user ((`product.limitPerUser && product.limitPerUser >= quantity)`) except for deluxe users (`security.isDeluxe(req)`)

However, it does **not** check if the quantity is a positive whole number.
```

This small mistake is costly. Users can freely input negative values for

> `quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100

Reviewer: changes the meaning slightly:

- attacker purchases `-100` units
- he needs to check out and use the `wallet` payment option 
- then (item cost * qty) gets deducted from his wallet - at this point, 100 times value of item is added to their wallet

Author: I think I've corrected the text to match your meeting in the [next comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945700345)

> This small mistake is costly. Users can freely input negative values for
> `quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100
> times the value of the item to their wallet.

Author: ```suggestion
This small mistake is costly. Users can freely input negative values for
`quantity`. In that case:

- Assume an attacker adds `-100` units to their basket
  - The attacker then checks out of their shopping cart, with the `wallet` payment option
- The Juice Shop adds 100 times the value of the item to the wallet
```

### Impact

Without such checks, this error could lead to substantial business losses.

### Prevention

In the following code sample, we include code that:

> - Checks if the quantity is greater than zero (`quantity > 0`)

Reviewer: Thanks, @mjang, looks really good.
 
I wanted to edit the specific code in the bracket to `quantity <= 0` to match the code fix. This is my mistake.

Author: Oh duh right! I should have seen it too.

```suggestion
- Throws an error for invalid quantities (`quantity <= 0`)
```

- Runs a boolean check for decimal integers (`Number.isInteger(quantity, 10)`)
- Returns an HTTP 400 error if the noted checks fail

```ts
async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
  const product = await QuantityModel.findOne({ where: { ProductId: id } })
  if (!product) {
    throw new Error('No such product found!')
  }

 // Check if the quantity is a positive decimal integer value, if not send a `400` error response
 if(!Number.isInteger(quantity, 10) || quantity <= 0) {
      res.status(400).json({ error: res.__('Invalid Quantity of items.') })
 }

  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
    if (product.quantity >= quantity) { // enough in stock?
      next()
    } else {
      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
    }
  } else {
    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
  }
}
```

## Best Practices

> 

Reviewer: Two more points I think I missed:

- Always check if the user input matches the right format.
- Always ensure that the user input fits the given business context
  - A user should only enter a number in a quantity field, but the user should not be able to enter negative values there.
  - There should be sensible minimum and maximum values for a given quantity.

Author: I love examples!

I've rewritten what you've suggested slightly, and added it in https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r946719506

- Assume that any input can be an attack
- Validate all user input before processing it
  - The [OWASP Input Validation Cheat Sheet
    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)

> that you use input validation as early as possible in the data flow,

Author: ```suggestion
    that you use input validation as early as possible in the data flow,
```

> - Assume that any input can be an attack
> - Validate all user input before processing it
>   - The [OWASP Input Validation Cheat Sheet
>     recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
>     that you use input validation as early as possible in the data flow,

Author: Fix punctuation for consistency

```suggestion
- Assume that any input can be an attack.
- Validate all user input before processing it.
  - The [OWASP Input Validation Cheat Sheet
    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
    that you use input validation as early as possible in the data flow.
```

> that you use input validation as early as possible in the data flow,
> - Include an **allow list** of valid inputs. Attackers can often bypass the deny
>   lists. For example, you could use the following to validate a US zip code:
> 
> ```ts
> // define the allow list regex for US Zip code
> const validZipPattern = /^\d{5}(-\d{4})?$/;
> // check if the user input matches our defined allow list regex
> if (!validZipPattern.test(userInput) {
>  		throw new Error("InvalidZipCode");
> }	
> ```
> 
> - Log all input validation failures. The information you get can help you detect
>   intrusions, and identify attackers who try to send invalid inputs. For this
> code sample, you can add the following logging statement: `logger.warn("Invalid zip code provided.")`:
> 
> ```ts
> // define the allow list regex for US Zip code
> const validZipPattern = /^\d{5}(-\d{4})?$/;
> // check if the user input matches our defined allow list regex
> if (!validZipPattern.test(userInput) {
>   // log the input validation failure
>   logger.warn("Invalid zip code provided.");
>   throw new Error("InvalidZipCode");
> }
> ```

Author: Revise formatting per Markdown requirements

````suggestion
that you use input validation as early as possible in the data flow,
- Include an **allow list** of valid inputs. Attackers can often bypass the deny
  lists. For example, you could use the following to validate a US zip code:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
   		throw new Error("InvalidZipCode");
  }	
  ```

- Log all input validation failures. The information you get can help you detect
  intrusions, and identify attackers who try to send invalid inputs. For this
  code sample, you can add the following logging statement: `logger.warn("Invalid zip code provided.")`:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
    // log the input validation failure
    logger.warn("Invalid zip code provided.");
    throw new Error("InvalidZipCode");
  }
  ```
````
